ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+    //        A pure static shape for both ngraph function generation and input tensor
+    //
+    //  {}
+    //        A unspecified shape, test_unary() will derive the actual shape from the size
+    //        of the input vector as 1D shape.
+    struct ArgShape
+    {
+        PartialShape dynamic_shape;
+        Shape static_shape;
+        bool initialzed;
+        ArgShape(const PartialShape& dynamic_shape, const Shape& static_shape)
+            : dynamic_shape(dynamic_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape(const Shape& static_shape)
+            : dynamic_shape(static_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape()
+            : dynamic_shape{}
+            , static_shape{}
+            , initialzed(false)
+        {
+        }
+    };
+
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& test_input,
+                    const std::vector<value_type>& test_expected,
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        if (!ashape.initialzed)
+        {
+            ashape = ArgShape(Shape({test_input.size()}));
+        }
+        // the size specified in static_shape may be bigger or smaller than the
+        // number of data contained in test_input & test_expected, here we
+        // sample test data repetitively to generate actual test data of required size.
+        auto static_size = shape_size(ashape.static_shape);
+        std::vector<value_type> input;
+        std::vector<value_type> expected;
+        for (size_t i = 0; i < static_size; i++)
+        {
+            input.push_back(test_input[i % test_input.size()]);
+            expected.push_back(test_expected[i % test_expected.size()]);
+        }
+
+        auto test_case = test::TestCase<TestEngine>(f);
+        test_case.template add_input<value_type>(input);
+        test_case.template add_expected_output<value_type>(ashape.static_shape, expected);
+        if (tol.using_bits)
+        {
+            if (tol.tolerance_bits >= 0)
+                test_case.run(tol.tolerance_bits);
+            else
+                test_case.run();
+        }
+        else
+        {
+            test_case.run_with_tolerance_as_fp(tol.tolerance_fp);
+        }
+    }
+
+    // overloading support passing reference function instead of expected results
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& input,
+                    value_type (*func)(value_type),
```

Comment:
Can this function pointer be changed to something that accepts a general callable? Also the name `func` is too vague, it should be named differently so that it explains what the executed callable actually produces.

Reply:
I tried, it's not work. compiler failed to deduce the argument type for the general callable when you pass an overloaded function name to it, for example, all most all functions ("exp","log","sin",...) from <cmath> have many overloads. 

On the other hand,  since value_type is already known, "value_type (*func)(value_type)" tells compiler the exact overload to use, so it works well.

Reply:
I will change func to get_expect

Reply:
Just one more request - please don't use such abbreviations. `get_expecteded_output` or `expected_output_generator` are not that long but it will be much easier to understand what they do.

Reply:
Agree